README.md
# A System Thing

Responsive, theme-based style props for building design systems in ~~React~~ JavaScript

[![Build Status][ci]][travis]
[![Version][]][npm]
![size][]
![MIT License][license]

[ci]: https://flat.badgen.net/travis/smuemd/systemthing
[version]: https://flat.badgen.net/npm/v/systemthing
[license]: https://flat.badgen.net/badge/license/MIT/blue
[size]: https://flat.badgen.net/bundlephobia/minzip/systemthing

[npm]: https://npmjs.com/package/systemthing
[travis]: https://travis-ci.org/smuemd/systemthing
[github]: https://github.com/smuemd/systemthing

```sh
npm i systemthing
```

This repository is a fork of the [styled-system](https://github.com/jxnblk/styled-system).

As with styled-system, this fork presents a collection of css-style yielding functions that support the development of responsive and theme-based design systems in JavaScript.

Unlike styled-system, this implementation is geared to also work with virtual dom libraries other than React.

## Features

This fork provides the same core functionality found in the [orignal styled-system libary](https://github.com/jxnblk/styled-system#features),

plus:
+ Works with any virtual DOM view layer (e. g. [Mithril](https://mithril.js.org/))
+ Works great in tandem with [bss](https://github.com/porsager/bss) and [stylething](https://github.com/smuemd/stylething).

minus:
+ Social proof, i. e. not used in [Rebass](https://rebassjs.org) et al.

### Table of Contents

- [Notable Differences](#notable-differences)
- [Usage](#usage)
- [Docs](#docs)
- [Related](#related)
<!-- TODO
- [Getting Started](#getting-started)
- [Further Reading](#further-reading)
-->

## Notable differences

While this implementation tries to stay as close as possible to [styled-system's original API](https://github.com/jxnblk/styled-system/blob/master/docs/api.md), the following changes were made:

### non-contextual theming

Style yielding functions can be called with a custom theme by passing a theme-object as a second argument.

```js
import { space, fontSize, color } from 'systemthing'

const theme = {
 breakpoints: [ '32em', '48em', '64em' ],
 space: [ 0, 6, 12, 18, 24 ],
 fontSizes: [ 12, 16, 18, 24, 36, 72 ],
 radii: [ 2, 4 ],
 colors: {
   blue: '#07c',
   green: '#1c0',
   gray: ['#ccc', '#555']
 }
}
                       
// aliases
theme.space.big = 64
theme.fontSizes.big = 128

const blue = color({ color: 'blue' }, theme) // yields { color: '07c' }
const big = fontSize({ fontSize: 'big' }, theme) // yields { fontSize: '128px' }
const butAlso = space({ padding: 'big', theme }) // yields { padding: '64px' }
```
Passing a theme as 2nd argument bypasses [React `context`](https://reactjs.org/docs/context.html) and allows this implementation to work in concert with any virtual DOM based view layer.

It is still possible however to also use `theme` as a prop on the first argument. Prop based themes take precedence. [styled-system's original theming mechanism](https://github.com/jxnblk/styled-system/blob/master/docs/theming.md) remains intact.

### No prop abbreviations

Mapping of popular style prop abbreviations to corresponding css properties (e. g. `mt` to `marginTop`) is delegated downstream. The only exceptions are `mx`, `my`, `px` and `py` within the `space` function. For there are no css equivalents to these props. 

### Default values

It is possible to directly instrument style yielding functions with default values as a fallback in case no corresponding style props are present at call time.

```js
import { space } from 'systemthing'

const spaceWithDefaults = space.withDefaults({ padding: '1em' })

const spaced = spaceWithDefaults({ foo: 'bar' }) // { padding: '1em' }
```
Responsive default styles may be useful for defining components that do not require expensive style re-computations down the line.

### Prop bundling

To minimize the number of downstream function calls, style props are bundled into a single function where possible.

```js
import { flexbox } from 'systemthing'

const flexing = flexbox({
  display: 'flex',
  justifyContent: 'center',
  flexDirection: 'row-reverse'
})
```
The following prop bundles are available:

#### 1. `backgrounds`
bundled props:
`background`,
`backgroundImage`,
`backgroundPosition`,
`backgroundRepeat`,
`backgroundSize`

#### 2. `borders`
bundled props:
`border`,
`borderTop`,
`borderRight`,
`borderBottom`,
`borderLeft`

#### 3. `color`
bundled props: 
`color`,
`backgroundColor`

#### 4. `direction`
bundled props:
`top`,
`right`,
`bottom`,
`left`

#### 5. `flexbox`
bundled props:
`display`,
`alignContent`,
`alignSelf`,
`justifyContent`,
`justifyItems`,
`justifySelf`,
`flex`,
`flexBasis`,
`flexDirection`,
`flexWrap`,
`order`

#### 6. `gridLayout`
bundled props: 
`gridArea`,
`gridAutoColumns`,
`gridAutoFlo`',
`gridAutoRows`,
`gridRow`,
`gridRowGap`,
`gridTemplateAreas`,
`gridTemplateColumns`,
`gridTemplateRows`

#### 7. `gridGaps`
bundled props:
`gridColumnGap`,
`gridGap`,
`gridRow`,
`gridRowGap`,

#### 8. `space`
bundled props: 
`margin`,
`marginTop`,
`marginRight`,
`marginBottom`,
`marginLeft`,
`mx`,
`my`,
`padding`,
`paddingTop`,
`paddingRight`,
`paddingBottom`,
`paddingLeft`,
`px`,
`py`

### No PropTypes
Facebook's [`prop-types`](https://github.com/facebook/prop-types) has been removed. Instead type declaration is delegated downstream. Users of this fork have to be prepared to properly type out components themselves, if necessary. This fork ships without external dependencies. 

### Width fit & fill
The `width` style yielding function accepts two additional prop values:
- `{ width: 'fit' }` - Make an element shrink wrap its content with `flex-basis`.
- `{ width: 'fill' }` - Make an element fill the remaining space. Distribute space evenly on multiple elements.

### low level changes

#### `get`
The `get` utility function accepts excactly two arguments `obj :: Object` and `path :: Array`. `get`'s `path` argument is not being destructured.

Hence, instead of writing:
```js
// wrong
let bingo = get({ path: { in: { obj: 'Bingo' } } }, 'path', 'in', 'obj')
```
you'd have to write:
```js
// correct
let bingo = get({ path: { in: { obj: 'Bingo' } } }, [ 'path', 'in', 'obj' ])
```
Secondly, dot notation in `path` strings is not supported out of the box. The following statement will fail to produce the desired result:
```js
// wont work
let bingo = get({ path: { in: { obj: 'Bingo' } } }, [ 'path.in.obj' ])
```
Note: Refactoring the `get` function apparently resulted in significant performance improvements for some style functions.

#### `num`
The `num` utility is renamed to **`isNum`**.

#### `px`
The `px` transformation utility is renamed to **`addPx`**.

#### `cloneFunc`
The `cloneFunc` utility has been removed together with the *`prop-types`* dependency.

#### `merge`
The `merge` utility is renamed to **`mergeStyles`**.

#### `compose`
The `compose` utility is renamed to **`composeStyleFns`**. (may be removed entirely, as it is only used in a benchmarking script right now)

#### `style`
The `style` utility function is renamed to **`createStyleFn`**.
The original function was refactored to also process arrays when passed as `prop` and `cssProptery` arguments. This enables the prop bundling functionality described above.

```js
import { util } from 'systemthing'

const custom = util.createStyleFn({
  prop: [ 'display', 'overflow', 'opacity' ]
})

const result = custom({ display: 'none', overflow: 'hidden' })
```

#### `mixed`
The `mixed` utility has been removed entirely.

## Usage

This implementation maintains most of [styled-system's original API](https://github.com/jxnblk/styled-system/blob/master/docs/api.md).

Here is an idiomatic usage example:
```js
// Usage in tandem with compatible css-in-js lib is recommended
// Example uses bss, but systemthing works with most other css-in-js libraries as well
import b from 'bss'
import { color, fontSize, util } from 'systemthing'

// custom theme
const theme = {
 breakpoints: [ '32em', '48em', '64em' ],
 space: [ 0, 6, 12, 18, 24 ],
 fontSizes: [ 12, 16, 18, 24, 36, 72 ],
 colors: {
   blue: '#07c',
   green: '#1c0',
   grays: [ '#ccc', '#555', '#ddd' ]
 }
}

// define style props
const props = {
  // responsive background-color (theme.colors)
  backgroundColor: [ 'blue', 'green', 'grays.1', 'grays.2' ],
  // responsive font-size of 24px, 36px or 72px (theme.fontSizes)
  fontSize: [ 3, 4, 5 ],
  // theme passed as prop
  theme
}

// pass style props into the style functions and merge their return values
const result = util.mergeStyles(color(props), fontSize(props))

// result is a responsively themed css style object that
// can then be processed downstream by many css-in-js libraries
const classy = b(result).class

document.body.classList.add(classy)
document.body.appendChild(
  document.createTextNode(
    'Responsively themed! (resize the window!)'))

```

CHANGELOG.md

# Changelog

## v1.74 2018-12-07 

- Style yielding functions can be called with a custom theme by passing a theme-object as a second argument.
- Mapping of popular style prop abbreviations to corresponding css properties (e. g. `mt` to `marginTop`) is delegated downstream. The only exceptions are `mx`, `my`, `px` and `py` within the `space` function.
- Directly instrument style yielding functions with default values as a fallback in case no corresponding style props are present at call time.
- `backgrounds`, `borders`, `color`, `direction`, `flexbox`, `gridLayout`, `gridGaps`, and `space` functions bundle multiple style props each.
- remove [`prop-types`](https://github.com/facebook/prop-types) dependency.
-  the `width` function now accepts 'fit' and 'fill' prop values
- refactor of `get` utility function to accept excactly two arguments `obj :: Object` and `path :: Array`. No more destructuring. Dot notation in `path` strings is no longer supported out of the box.
- `num` utility renamed to **`isNum`**
- `px` transformation utility renamed to **`addPx`**
- removed `cloneFunc` utility
- `merge` utility renamed to **`mergeStyles`**
- `compose` utility is renamed to **`composeStyleFns`**
- `style` utility function renamed to **`createStyleFn`**
- `style` utility function refactored to now also accept arrays when passed as `prop` and `cssProptery` arguments.
- Removed `mixed` utility function

## v3.1.10 2018-10-10

- Convert numbers to pixels with keyed space values

## v3.1.9 2018-10-10

- Allow `getTheme` to return falsy values

## v3.1.8 2018-10-10

- Use Babel runtime

## v3.1.7 2018-10-10

- Fix for `breakpoints` numbers not being converted to pixels

## v3.1.6 2018-09-23

- Fix `.npmignore`

## v3.1.5 2018-09-23

- Update docs

## v3.1.4 2018-09-19

- Fix for passing falsy value as first item in array values

## v3.1.3 2018-09-14

- Add `module` field to package.json

## v3.1.2 2018-09-14

- Fix missing `styles` object keys

## v3.1.1 2018-09-14

- Patch to debug broken release

## v3.1.0 2018-09-11

- Upgrade to Babel 7

## v3.0.0

- Performance rewrite
- Simplified API
- Faster
- Smaller

### Added

- `variant`
- `compose`
- `css`

### Changed

- `style` utility is now responsive by default
- All built-in style functions now support responsive array values
- `backgroundImage` no longer wraps values with `url()`
- `buttonStyle` prop has been renamed to `variant`
- The `style` utility `getter` argument has beend renamed to `transformValue`
- `util.get` no longer supports a fallback argument

### Removed

- `propTypes[prop].meta` (temporarily removed)
- `responsiveStyle` use `style` instead
- `complexStyle` use `variant` instead
- `complexStyle` (`variant`) boolean props are no longer supported
- `pseudoStyle`
- `hover`
- `focus`
- `active`
- `disabled`
- `fontSize` and `width` no longer support shorthand props
- `alias` option
- `numberToPx` option (pass `util.px` to the `transformValue` argument instead)
- `util.media`
- `util.neg`
- `util.arr`
- `util.getWidth`
- `util.mq`
- `util.fallbackTheme`
- `util.breaks`
- `util.dec`
- `util.getValue`

## 2.3.0

### Added

- Metadata for supporting system-docs package

## 2.2.0

### Added

- CSS Grid utilities

## 2.1.0

### Changed

- Reorganized code base

## 2.0.0

### Added

- More utilities for CSS properties
- `fontFamily`
- `display`
- `minWidth`
- `height`
- `maxHeight`
- `minHeight`
- `size` (width & height)
- `alignContent`
- `order`
- `borders` utility that normalizes the API with other styled-system utilities
- `backgroundImage`
- `backgroundSize`
- `backgroundPosition`
- `zIndex`
- `top`
- `right`
- `bottom`
- `left`

### Changed

- Uses verbose naming convention for `textAlign`, `alignItems`, `justifyContent`, `flexWrap`, and other CSS properties
- `style`, `responsiveStyle`, and `pseudoStyle` utilities now accept `getter` and `getters` arguments for converting values
- Rewritten with ES module syntax
- DRYer code base for utilities like `fontSize`, `width`, and `color`
- The repo has been reorganized as a monorepo
- The `flexWrap` utility now expects a string value for its prop (previously a boolean)
- The `borderTop`, `borderRight`, `borderBottom`, and `borderLeft` props no longer accept boolean props, but follow the same convention as other utilities
- The default spacing scale has changed to `[ 0, 4, 8, 16, 32, 64, 128, 256, 512 ]` (this only affects usage without a custom `theme.space` provided
- Number values for breakpoints are now converted to px instead of ems to better match the rest of the library
- The `theme` utility has been renamed to `themeGet` (alternative names under consideration)
- The `pseudoStyle` arguments API has changed to more closely align with `style` and `responsiveStyle`

### Removed

- The `cleanElement` utility is now a separate package, removing the `react` dependency from the core library
- The `borderWidth` utility has been replaced with the `borders` utility
- The `theme` utility has been renamed to `themeGet` (alternative names under consideration)
- `removeProps` utility
- `idx` utility
- Legacy `responsiveStyle` arguments API




lib/core.js
import { addPx, createStyleFn, getWidth } from './util'

// space

export const space = createStyleFn({
  prop: [
    'margin',
    'marginTop',
    'marginRight',
    'marginBottom',
    'marginLeft',
    'mx',
    'my',
    'padding',
    'paddingTop',
    'paddingRight',
    'paddingBottom',
    'paddingLeft',
    'px',
    'py'
  ],

  cssProperty: [
    'margin',
    'marginTop',
    'marginRight',
    'marginBottom',
    'marginLeft',
    [ 'marginRight', 'marginLeft' ],
    [ 'marginTop', 'marginBottom' ],

    'padding',
    'paddingTop',
    'paddingRight',
    'paddingBottom',
    'paddingLeft',
    [ 'paddingRight', 'paddingLeft' ],
    [ 'paddingTop', 'paddingBottom' ]
  ],

  key: 'space',

  transformValue: addPx,

  scale: [ 0, 4, 8, 16, 32, 64, 128, 256, 512 ]
})

// width

function intercept (input) {
  const list = Object.keys(input || {})
  let res = {} // Object.assign({}, input)
  for (let i = 0; i < list.length; i++) {
    const k = list[i]
    const ik = input[k]
    if (ik === 'fill') {
      res['flex'] = '1 1 0%'
      res['flexBasis'] = '0%'
    } else if (ik === 'fit') {
      res['flexBasis'] = 'auto'
    } else if (typeof ik === 'object') res[k] = intercept(ik)
    else res[k] = ik
  }
  return res
}

export const widthSimple = createStyleFn({
  prop: 'width',
  transformValue: getWidth
})

export const width = (attrs, opts = {}) => {
  const val = attrs.width
  const fitOrFill = Array.isArray(val)
    ? (val.indexOf('fit') >= 0 || val.indexOf('fill') >= 0) // val.some(x => x === 'fit' || x === 'fill')
    : (val === 'fill' || val === 'fit')
  return fitOrFill ? intercept(widthSimple(attrs, opts)) : widthSimple(attrs, opts)
}

// fontSize

export const fontSize = createStyleFn({
  prop: 'fontSize',
  key: 'fontSizes',
  transformValue: addPx,
  scale: [
    12,
    14,
    16,
    20,
    24,
    32,
    48,
    64,
    72
  ]
})

// color

export const textColor = createStyleFn({
  prop: 'color',
  key: 'colors'
})

export const bgColor = createStyleFn({
  prop: 'backgroundColor',
  cssProperty: 'backgroundColor',
  key: 'colors'
})

export const color = createStyleFn({
  prop: [ 'color', 'backgroundColor' ],
  // cssProperty: [ 'color', 'backgroundColor' ],
  key: 'colors'
})
